Alternate stylesheets are not present in document.styleSheets

rniwa opened the pull request:
#### a39b520b8e0ca63489db7a45fe19181a33fac903
<pre>
Alternate stylesheets are not present in document.styleSheets
<a href="https://bugs.webkit.org/show_bug.cgi?id=56932">https://bugs.webkit.org/show_bug.cgi?id=56932</a>

Reviewed by Cameron McCormack.

Align the behavior of WebKit with the spec and that of other browsers by always enumerating list of
stylesheets regardless of whether if title is specified or it is an alternate stylesheet.

Also fixed the bug that StyleSheet.title returns an empty string instead of null.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/stylesheet-title-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/ttwf-cssom-doc-ext-load-tree-order-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/wincairo/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:

* Source/WebCore/css/CSSStyleSheet.h:
(WebCore::XSLStyleSheet::title): Return null if m_title is an empty string.
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectActiveStyleSheets): This function now returns two vector&apos;s containing
the list of active stylesheets as well as ones for Document/ShadowRoot&apos;s styleSheets attribute.
(WebCore::Style::Scope::updateActiveStyleSheets):
* Source/WebCore/style/StyleScope.h:
(WebCore::Style::Scope::ActiveStyleSheetCollection):
* Source/WebCore/xml/XSLStyleSheet.h:
(WebCore::XSLStyleSheet::title): Return null.

Canonical link: <a href="https://commits.webkit.org/252781@main">https://commits.webkit.org/252781@main</a>
</pre>

aproskuryakov commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -338,20 +338,25 @@ Vector<Ref<ProcessingInstruction>> Scope::collectXSLTransforms()
 #endif
 
-void Scope::collectActiveStyleSheets(Vector<RefPtr<StyleSheet>>& sheets)
+auto Scope::collectActiveStyleSheets() -> ActiveStyleSheetCollection
```
Do we now prefer this style in WebKit? I thought that it was an ugly exception only needed in rare template programming situations, not a generally preferred style.

rniwa replied:
No that we prefer but this avoids having to repeat `Style::Scope::`.

heycam commented on Source/WebCore/css/CSSStyleSheet.h:
```diff
@@ -60,7 +60,7 @@ class CSSStyleSheet final : public StyleSheet {
     MediaList* media() const final;
     String href() const final;
-    String title() const final { return m_title; }
+    String title() const final { return !m_title.isEmpty() ? m_title : String { }; }
```
Nit: I guess I'd write `String()` to construct a null string rather than use the brace form.

nt1m replied:
Don't we have nullString() for null strings? cc @cdumez

rniwa replied:
`nullString()` exists but these two are interchangeable.

rniwa committed:
Alternate stylesheets are not present in document.styleSheets
https://bugs.webkit.org/show_bug.cgi?id=56932

Reviewed by Cameron McCormack.

Align the behavior of WebKit with the spec and that of other browsers by always enumerating list of
stylesheets regardless of whether if title is specified or it is an alternate stylesheet.

Also fixed the bug that StyleSheet.title returns an empty string instead of null.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/stylesheet-title-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/ttwf-cssom-doc-ext-load-tree-order-expected.txt:
* LayoutTests/platform/glib/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/ios/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/mac/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/win/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:
* LayoutTests/platform/wincairo/tables/mozilla_expected_failures/bugs/bug92868_1-expected.txt:

* Source/WebCore/css/CSSStyleSheet.h:
(WebCore::XSLStyleSheet::title): Return null if m_title is an empty string.
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::collectActiveStyleSheets): This function now returns two vector's containing
the list of active stylesheets as well as ones for Document/ShadowRoot's styleSheets attribute.
(WebCore::Style::Scope::updateActiveStyleSheets):
* Source/WebCore/style/StyleScope.h:
(WebCore::Style::Scope::ActiveStyleSheetCollection):
* Source/WebCore/xml/XSLStyleSheet.h:
(WebCore::XSLStyleSheet::title): Return null.

Canonical link: https://commits.webkit.org/252781@main
